coralproject/talk
# [CORL-2473]: member role

## TODO
- [x] create memberShip scopes field
- [x] create UI for assigning site scopes to MEMBER role (refactor SiteModerator*.tsx -> SiteRole*.tsx)
- [x] update resolvers for `UserMembershipScopes`
- [x] add mutation for `UserMembershipScopes`
- [x] update resolvers to handle `Settings.staff` -> `Settings.badges` rename
- [x] refactor StaffTagContainer name
- [x] add config for default MEMBER badge to schema (in config model, and update config input/payload)
- [x] modify user role mutation + user upgrade mutation to handle commenter -> member, with appropriate permissions checks
- [x] add UI for configuring default member badge
- [x] discover what is needed migration wise

## Decision log

## What does this PR do?
This PR creates a new role, `MEMBER` that currently doesn't grant any privileges but: 
1. causes a user's comments to appear with a badge and
2. can be org wide or scoped to specific sites, similar to a Site Moderator

The default value for the badge is "MEMBER", but this value can be configured by admins (similar to staff badges).

Overview of changes:
- Naming: things named for Site Moderators have been either made more general (Site Moderator -> Site Role) to accommodate the addition of the non-moderator, but site specific "Member role", or they have been made more specific (PromoteUser -> PromoteModerator) to distinguish between moderator specific actions and related, but distinct, member specific actions.
- The SiteModeratorModal + Actions component have been refactored to accommodate both roles.

## What changes to the GraphQL/Database Schema does this PR introduce?
This introduces the `MEMBER` value to the UserRole enum, as well as the `UserMembershipScopes` type, and adds the nullable field `membershipScopes: UserMembershipScopes` to the `User` type. This also renames `Settings.staff: StaffConfiguration ` -> `Settings.badges: BadgeConfiguration` (along with related Input types).

## Does this PR introduce any new environment variables or feature flags? 
No.

## If any indexes were added, were they added to `INDEXES.md`?
No.

## How do I test this PR?
- As an admin, change some users roles to "MEMBER" on any site in the org
-  see that those users receive the MEMBER badge on the sites on which they have been granted MEMBER status.
- as a site mod, see that you can add/remove your sites to a member (or to any user with role < moderator), but no others
- as an admin, under Confguration -> General -> Member Badges, update the badge text for members and see that their comments receive the new text as a badge
- IMPORTANT: I've tested this flow manually and am confident that it all edge cases have been handled, but make sure to take note of the previous values for Staff/Moderator/Admin badge texts are before deploying this branch, and confirm that they are preserved after deploying it, since the name of this Settings field has changed.
 
 
## How do we deploy this PR?
Outdated/cached frontend JS will throw an error when it first receives a user with role "MEMBER" from the backend.
